Issue/200 authorization

iperdomo opened the pull request:
* Filter changes by authorization
* Removes `flumenfly` related code

iperdomo committed:
[#200] Remove /check_permission endpoint

iperdomo committed:
[#200] Filter changes by authorized forms

* formChanged|formDeleted will show only data the user has access to

iperdomo committed:
[#200] Move ds/with-datastore outside list/filter calls

* We expect that RemoteAPI to be installed once per thread (request),
instead of installing/uninstalling several times

iperdomo committed:
[#200] Filter forms/form-instances/surveys/data-points by authorization

dlebrero commented on api/src/clojure/org/akvo/flow_api/unilog/unilog.clj:
```diff
@@ -104,32 +105,32 @@
-                   set)
-   :form-deleted form-deleted
+   :form-changed (->> form-updated (keep form-id->form) set)
+   :form-deleted (set/intersection authorized-forms form-deleted)
```
A deleted form will never ever be in the authorised-forms, so the :form-deleted will always always be the empty set.

Deleted forms cannot be authorised. Should be made clear in the API doc that any delete may contain objects that the client does not know about.

Note that the end to end tests are lying here, as they are not properly simulating the whole flow, as they do not delete the objects in the datastore even if they put a "delete" message in the "fake" unilog DB.

iperdomo replied:
> A deleted form will never ever be in the authorised-forms, so the :form-deleted will always always be the empty set.

You're right and I thought about it. When we find a `form-deleted` the form is already gone, so looking for authorized forms will not give us that id. :thinking:

iperdomo replied:
> Note that the end to end tests are lying here, as they are not properly simulating the whole flow, as they do not delete the objects in the datastore even if they put a "delete" message in the "fake" unilog DB.

Correct. You're right.

dlebrero commented on api/src/clojure/org/akvo/flow_api/unilog/unilog.clj:
```diff
@@ -104,32 +105,32 @@
+                                        (when-let [form (get form-id->form form-id)]
+                                          {:form form
+                                           :form-instance-ids (set (map :id form-instance))})))
+                                set) ;; the mapping form-id->form filter instances
```
This comment sounds like the code is not clear. Please fix the code and remove the comment

dlebrero commented on api/src/clojure/org/akvo/flow_api/unilog/unilog.clj:
```diff
@@ -104,32 +105,32 @@
+                                               survey-changed
+                                               survey-deleted]}
+                                      form-id->form
+                                      authorized-forms
```
No see the need of both the authorised-forms and the form-id->form. As it is `authorised-forms == (keys form-id->form)`

dlebrero commented on api/src/clojure/org/akvo/flow_api/unilog/unilog.clj:
```diff
@@ -104,32 +105,32 @@
-   :survey-changed survey-changed
-   :survey-deleted survey-deleted})
+   :survey-changed (set/intersection authorized-surveys survey-changed)
+   :survey-deleted (set/intersection authorized-surveys survey-deleted)})
```
Same comment as above regarding deletes. Survey deletes cannot be authz.

dlebrero commented on api/src/clojure/org/akvo/flow_api/unilog/unilog.clj:
```diff
@@ -142,26 +143,37 @@
-                                         {:auto-commit? false :fetch-size 300}))
+          events (process-new-events (read-event-log db offset))
+          user-id (user/id-by-email-or-throw-error remote-api instance-id email)
+          authorized-forms (set (map ds/id (su/list-forms user-id)))
```
Why not load just the required `(:forms-to-load events)`? I expect that in each batch there will be a little number of different forms to load.

dlebrero commented on api/src/clojure/org/akvo/flow_api/unilog/unilog.clj:
```diff
@@ -142,26 +143,37 @@
-                                (:forms-to-load events))
-          events-2 (after-forms-loaded events form-id->form) ;; TODO: get form definition from cache
+                                authorized-forms-to-load)
+          authorized-surveys (set (map ds/id (su/list* user-id)))
```
Why not load just the required surveys? I expect that in each batch there will be a little number of different surveys to load.

dlebrero commented on api/src/clojure/org/akvo/flow_api/unilog/unilog.clj:
```diff
@@ -142,26 +143,37 @@
-          data-point-changed (doall (data-point/by-ids ds (:data-point-changed events-2)))]
-      (assoc events-2 :form-instance-changed form-instances :data-point-changed data-point-changed))))
+                             (:form-instances-to-load authorized-events)))
+          data-point-changed (doall (data-point/by-ids ds (map :id (:data-point-changed authorized-events))))
```
The result from `filter-events-by-authorization` should be manipulated as little as possible, because `filter-events-by-authorization` is properly unit tested, while this code is not. Hence I would suggest that `filter-events-by-authorization` should just return the `id` if that is the only thing needed.

dlebrero commented on api/src/clojure/org/akvo/flow_api/unilog/unilog.clj:
```diff
@@ -86,9 +87,9 @@
                                                     (distinct (keep ::form-instance-changed final)))))
                data-point-deleted (set (keep ::data-point-deleted final))
-               data-point-changed (apply disj (set (keep ::data-point-changed final)) data-point-deleted)
+               data-point-changed (remove #(data-point-deleted (:id %)) (set (keep ::data-point-changed final)))
```
Why not remove data-points for surveys that we know have been deleted in this batch? Similar to what we are doing a few lines above (https://github.com/akvo/akvo-flow-api/pull/206/files#diff-0fa2688d1c94f616fb92da2720ce12f4R85) with the form-instances.

Maybe this comment is more general. Do we need to promise that you will not get an update for a child object of an object deleted? 
Because maybe if we promise that, then we also need to remove updates of forms of deleted surveys, form instances of deleted surveys, form instances of deleted data points, ... 
Is this possible? Is it easy? Is 100% accurate?

dlebrero replied:
I see that we did "just" the delete-survey removes data-points updates. 

What are your thoughts about the delete-survey should delete form-instances-updates?

dlebrero replied:
I just realised that if a survey is deleted, then when we go to load the forms due to a form-instance-update or a form-update, the forms should not be in the Google Datastore anymore. Same goes for a form deleted and a form-instance-update.

This holds true as long as we do not cache the calls to load surveys/forms.

I do not know enough about datapoints.

dlebrero commented on api/src/clojure/org/akvo/flow_api/unilog/unilog.clj:
```diff
@@ -104,32 +105,32 @@
+                                          {:form form
+                                           :form-instance-ids (set (map :id form-instance))})))
+                                set) ;; the mapping form-id->form filter instances
+   :data-point-changed (filter #(authorized-surveys (:survey-id %)) data-point-changed)
```
Maybe use `comp` instead of anon fn

dlebrero commented on api/src/clojure/org/akvo/flow_api/unilog/unilog.clj:
```diff
@@ -142,26 +143,37 @@
-      (assoc events-2 :form-instance-changed form-instances :data-point-changed data-point-changed))))
+                             (:form-instances-to-load authorized-events)))
+          data-point-changed (doall (data-point/by-ids ds (map :id (:data-point-changed authorized-events))))
+          survey-changed (doall (su/by-ids ds (:survey-changed authorized-events)))]
```
`authorized-surveys` already goes and loads all the data required here. 

Maybe do something similar as with forms, where the `process-new-events` function returns the forms that need to be loaded, and later the `filter-events-by-authorization` uses the `form-id->form` to return the full forms, plus to do the authz on form-instances.

dlebrero reviewed (changes requested):
Good. Almost there!

iperdomo committed:
[#200] Refactor the way of handling authorization

- We can't authorize deletes (as the main object on which we authorize may be gone)
- Provide candidates for forms and surveys ids when trying to list-and-authorize

dlebrero commented on api/src/clojure/org/akvo/flow_api/datastore/survey.clj:
```diff
@@ -1,16 +1,23 @@
-        all-surveys (.listAll survey-dao)]
-    (.filterByUserAuthorizationObjectId survey-dao all-surveys user-id)))
+(defn list*
+  ([user-id survey-ids]
```
It is going to be very confusing if by mistake we pass a null here and end up getting all surveys. Same goes for the forms.

The duplication removal is not worth the possible confusion and overload of the survey-ids parameter meaning (null vs empty vs not-empty).

iperdomo replied:
> It is going to be very confusing if by mistake we pass a null here and end up getting all surveys. Same goes for the forms.

We **always filter** by authorization, the possibility of providing a set if ids (forms or surveys) is an _optimization_ to get fewer objects and and filter those by authorization. 

A call will never get _all surveys_ if the user is not authorized.

Does it make sense?

dlebrero commented on api/src/clojure/org/akvo/flow_api/unilog/unilog.clj:
```diff
@@ -115,22 +115,24 @@
-   :data-point-changed (filter #(authorized-surveys (:survey-id %)) data-point-changed)
+                                set)
+   :data-point-changed (->> data-point-changed
+                            (filter (comp (set (keys survey-id->survey)) :survey-id))
```
`(set (keys survey-id->survey))` == `survey-id->survey` for this particular case.

A map is a fn.

dlebrero commented on api/src/clojure/org/akvo/flow_api/unilog/unilog.clj:
```diff
@@ -154,22 +156,25 @@
-          data-point-changed (doall (data-point/by-ids ds (map :id (:data-point-changed authorized-events))))
-          survey-changed (doall (su/by-ids ds (:survey-changed authorized-events)))]
+          data-point-changed (doall (data-point/by-ids ds (:data-point-changed authorized-events)))
+          survey-changed (vals survey-id->survey)]
```
Should use the result of `filter-events-by-authorization`, instead of doing the calculation here again

dlebrero commented on api/src/clojure/org/akvo/flow_api/unilog/unilog.clj:
```diff
@@ -154,22 +156,25 @@
-          authorized-surveys (set (map ds/id (su/list* user-id)))
-          authorized-events (filter-events-by-authorization events form-id->form authorized-forms authorized-surveys)
+                                authorized-forms)
+          authorized-surveys (su/list* user-id (:survey-changed events))
```
There is no `:survey-changed` key on the `process-new-events` result, but a `::survey-changed`, so  `authorized-surveys` always contains all the surveys the user can see. 

Surveys+Datapoints should follow a flow similar to the Form+FormInstances: `process-new-events` calculates the surveys to be loaded/permissioned, `filter-events-by-authorization` does the filtering/enriching logic.

Test happens to pass because **by mistake we pass a null here and end up getting all surveys**.

iperdomo committed:
[#200] Fix authorization logic based on last review

- Get rid of using `nil` value when loading forms and surveys. This
  particular one masked a different logic bug. By using explicit ids
  when fetching objects we harden the function "contract".

iperdomo committed:
[#200] Removes the possibility of passing nil as parameter

dlebrero commented:
Merged.  Just left one comment open for your thoughts and maybe some future work.
